src/group_theory/marking.lean

```diff
@@ -0,0 +1,142 @@
+
+namespace group_marking
+
+@[to_additive]
+instance : monoid_hom_class (group_marking G S) (free_group S) G :=
+{ coe := λ f, f.to_fun,
+  coe_injective' := λ f g h, by cases f; cases g; congr',
+  map_mul := λ f, f.map_mul',
+  map_one := λ f, f.map_one' }
+
+@[to_additive]
+lemma map_surjective (m : group_marking G S) : surjective m := m.to_fun_surjective
+
+end group_marking
+
+/-- The trivial group marking. -/
+@[to_additive "The trivial additive group marking."]
+def group_marking.refl : group_marking G G :=
+{ to_fun_surjective := λ x, ⟨free_group.of x, free_group.lift.of⟩,
+  ..free_group.lift id }
+
+@[to_additive]
+instance : inhabited (group_marking G G) := ⟨group_marking.refl⟩
+
+variables {G S} {m : group_marking G S}
+
+/-- A type synonym of `G`, tagged with a group marking. -/
+@[derive group, nolint unused_arguments,
+to_additive "A type synonym of `G`, tagged with an additive group marking."]
+def marked_group (m : group_marking G S) : Type* := G
+
+attribute [to_additive] marked_group.group
+
+/-- "Identity" isomorphism between `G` and a group marking of it. -/
+@[to_additive "\"Identity\" isomorphism between `G` and an additive group marking of it."]
+def to_marked_group : G ≃* marked_group m := mul_equiv.refl _
+
+/-- "Identity" isomorphism between a group marking of `G` and itself. -/
+@[to_additive "\"Identity\" isomorphism between an additive group marking of `G` and itself."]
+def of_marked_group : marked_group m ≃* G := mul_equiv.refl _
+
+@[simp, to_additive] lemma to_marked_group_symm_eq :
+  (to_marked_group : G ≃* marked_group m).symm = of_marked_group := rfl
+@[simp, to_additive] lemma of_marked_group_symm_eq :
+  (of_marked_group : marked_group m ≃* G).symm = to_marked_group := rfl
+@[simp, to_additive] lemma to_marked_group_of_marked_group (a) :
+  to_marked_group (of_marked_group (a : marked_group m)) = a := rfl
+@[simp, to_additive] lemma of_marked_group_to_marked_group (a) :
+  of_marked_group (to_marked_group a : marked_group m) = a := rfl
+@[simp, to_additive] lemma to_marked_group_inj {a b} :
+  (to_marked_group a : marked_group m) = to_marked_group b ↔ a = b := iff.rfl
+@[simp, to_additive] lemma of_marked_group_inj {a b : marked_group m} :
+  of_marked_group a = of_marked_group b ↔ a = b := iff.rfl
+
+variables (α : Type*)
+
+@[to_additive] instance [inhabited G] : inhabited (marked_group m) := ‹inhabited G›
+@[to_additive] instance marked_group.has_smul [has_smul G α] : has_smul (marked_group m) α :=
+‹has_smul G α›
+@[to_additive] instance marked_group.mul_action [mul_action G α] : mul_action (marked_group m) α :=
+‹mul_action G α›
+
+@[simp, to_additive] lemma to_marked_group_smul (g : G) (x : α) [has_smul G α] :
+  (to_marked_group g : marked_group m) • x = g • x := rfl
+@[simp, to_additive] lemma of_marked_group_smul (g : marked_group m) (x : α) [has_smul G α] :
+  of_marked_group g • x = g • x := rfl
+
+@[to_additive]
+private lemma mul_aux [decidable_eq S] (x : marked_group m) :
```

Comment:
If you want to keep this one private, then you should provide API for the norm before the end of the file, because otherwise it will be basically unusable. Or remove the `private` attribute.